### PR TITLE
Dev train

### PR DESCRIPTION
**Pull request by hstko**
making train.py and defining loss function for CVAE.

### CODE REVIEW

File: cvae/model.py [hunk 1]
```diff
@@ -50,28 +51,61 @@ def build_cvae_mlp(self, kl_weight=1.0):
...
+    def _vae_loss(self, x, x_decoded_mean):
+        '''
+        loss function for VAE
+
+        :param x: keras tensor object. target vector to be reconstructed.
+        :param x_decoded_mean: keras tensor object. output of decoder.
+        :return: loss
+        '''
+        ent_loss = self.input_dim * binary_crossentropy(x, x_decoded_mean)
+        kl_loss = - 0.5 * K.sum(1 + self._z_log_sigma - K.square(self._z_mean) - K.exp(self._z_log_sigma), axis=-1)
```
Comment (hsmtknj) [inline cvae/model.py:99]: ent_loss は Reconstruction error で
kl_loss は KL Divergence ででええんんかな？

復元誤差ってのは一般的に binary_crossentropyってので計算できるもんなん？
ここあんまり理解できてないです．すみません．

File: cvae/train.py [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+    :return: tuple of models, trained models (CVAE model, encoder model, decoder model)
+    '''
+
+    if not os.path.exists(dir_name):
+        os.makedirs(dir_name)
+
+    x_train, y_train, x_test, y_test, input_dim, n_classes = data
+
+    epochs = 20
+    batch_size = 100
```
Comment (hsmtknj) [inline cvae/train.py:49]: ここのepochsとかbatch_sizeってどうやって決めるん？
やっぱなにかしらをプロットしてみて決めたりするん？

File: cvae/train.py [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+                   epochs=epochs,
+                   batch_size=batch_size,
+                   validation_data=([x_test, y_test], x_test))
+
+    cvae_model.save_weights('{}/{}_.h5'.format(dir_name, model_name))
+
+    return cvae_model, encoder, decoder
+
+
+def plot_manifold(generator, dir_name='results/manifold'):
```
Comment (hsmtknj) [inline cvae/train.py:66]: これは結果可視化用の関数？

**Review by hsmtknj — APPROVED** [review]
approve

File: cvae/model.py (hunk 1 above)
Comment (hstko) [reply cvae/model.py:99]: あってます！
復元誤差はVAEの場合，カテゴリ分布を仮定するなら，binary crossentropyを使うって感じかなー．０から１の範囲でそれぞれのピクセルについて独立に確率を考えるって時はこれでいいと思うけど，一応，２乗誤差とかでも実装上は問題ない．（理論に基づいてはいないけれども・・・）

File: cvae/train.py (hunk 2 above)
Comment (hstko) [reply cvae/train.py:49]: そうやね！ エポック数とかは特に許容できる学習時間次第やと思う．
batch size とか学習率は色々研究はされてるけれども，問題依存なことも多い・・・

File: cvae/train.py (hunk 3 above)
Comment (hstko) [reply cvae/train.py:66]: その通り！
潜在変数空間に各データがどこに埋め込まれてるかを可視化してる！

File: cvae/model.py [hunk 4]
```diff
@@ -50,28 +51,61 @@ def build_cvae_mlp(self, kl_weight=1.0):
...
-        cvae_model = Model([x, y], dec_out)
-        encoder_model = Model([x, y], z_mean)
+        z = Input(shape=(self.z_dim,))
+        gen_merged = Concatenate(name='gen_merged')([z, y])
+        gen_hid = dec_dense(gen_merged)
+        gen_out = dec_out(gen_hid)
+
+        generator_model = Model([z, y], gen_out)
+
+        return cvae_model, encoder_model, generator_model
```
Comment (takami-sris) [inline cvae/model.py:88]: encoder_model, generator_modelの変数名を統一させたほうがいいかなと思いました。
あえて変数名を変更しているのなら問題ありませんすみません。

File: cvae/train.py [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+        os.makedirs(dir_name)
+
+    n = 10
+    digit_size = 28
+
+    figure = np.zeros((digit_size * n, digit_size * (n + 1)))
+
+    for i in range(n):
+        x = x_test[i]
+        y = y_test[i]
```
Comment (takami-sris) [inline cvae/train.py:138]: テストデータって10個しかないわけではないですよね？
とりあえずはじめの10個とってくかって感じですかね。

**Review by takami-sris — APPROVED** [review]
みましたー。

File: cvae/model.py (hunk 4 above)
Comment (hstko) [reply cvae/model.py:88]: 直しますすみません。

File: cvae/train.py (hunk 5 above)
Comment (hstko) [reply cvae/train.py:138]: コード的には良くないですすみません直します。
